sql/2024/04/image-formats.sql

```diff
@@ -0,0 +1,143 @@
+# HTTP Archive query to get median image size per image type for LCP images.
+#
+# WPP Research, Copyright 2024 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# See https://github.com/GoogleChromeLabs/wpp-research/pull/97
+
+DECLARE
+  DATE_TO_QUERY DATE DEFAULT '2024-03-01';
+
+CREATE TEMPORARY FUNCTION
+  IS_CMS(technologies ARRAY<STRUCT<technology STRING,
+                                   categories ARRAY<STRING>,
+                                   info ARRAY<STRING>>>,
+         cms STRING,
+         version STRING)
+  RETURNS BOOL AS ( EXISTS(
+  SELECT
+    *
+  FROM
+    UNNEST(technologies) AS technology,
+    UNNEST(technology.info) AS info
+  WHERE
+      technology.technology = cms
+    AND ( version = ""
+    OR ENDS_WITH(version, ".x")
+            AND (STARTS_WITH(info, RTRIM(version, "x"))
+        OR info = RTRIM(version, ".x"))
+    OR info = version ) ) );
+
+CREATE TEMPORARY FUNCTION
+  IS_IMAGE (summary STRING)
+  RETURNS BOOLEAN AS (STARTS_WITH(LOWER(CAST(JSON_EXTRACT_SCALAR(summary, "$.mimeType") AS STRING)), 'image/'));
+
+CREATE TEMPORARY FUNCTION
+  GET_LCP_ELEMENT_ATTRIBUTE(custom_metrics STRING,
+                            attribute STRING)
+  RETURNS STRING AS ( (
+  SELECT
+    JSON_EXTRACT_SCALAR(attr, "$.value") AS v
+  FROM
+    UNNEST(JSON_EXTRACT_ARRAY(JSON_EXTRACT(custom_metrics, '$.performance.lcp_elem_stats.attributes'))) AS attr
+  WHERE
+      JSON_EXTRACT_SCALAR(attr, "$.name") = attribute
+  LIMIT
+    1 ) );
+
+CREATE TEMPORARY FUNCTION
+  GET_LCP_IMAGE_ATTRIBUTE_VALUE(custom_metrics STRING,
+                                attribute STRING)
+  RETURNS STRING AS ( (
+  SELECT
+    JSON_EXTRACT_SCALAR(attributes, CONCAT("$.", attribute)) AS v
+  FROM
+    UNNEST(JSON_EXTRACT_ARRAY(custom_metrics, '$.Images')) AS attributes
+  WHERE
+    JSON_EXTRACT_SCALAR(attributes, CONCAT("$.", attribute)) IS NOT NULL
+    AND JSON_EXTRACT_SCALAR(attributes, CONCAT("$.", 'url')) = GET_LCP_ELEMENT_ATTRIBUTE(custom_metrics,
+                                                                                         'url')
```

Comment:
See my other comment. The `Images` data in `custom_metrics` provides some image specific details, e.g. `url` is the actual image loaded, and it has things like `naturalWidth` and `naturalHeight`. But the `performance.lcp_elem_stats.attributes` data literally only has the regular HTML attributes.

So at a minimum we need to do:
```suggestion
    AND JSON_EXTRACT_SCALAR(attributes, CONCAT("$.", 'url')) = GET_LCP_ELEMENT_ATTRIBUTE(custom_metrics,
                                                                                         'src')
```

Another idea worth trying would be to make the comparison a bit less strict. For example I just ran a test query for just the `pagesWithLcpImages` sub query, limiting results to `https://wordpress.org/` and including some more raw data to take a closer look. On that instance the `url` in `Images` includes some extra query parameters, while the `src` attribute doesn't. So potentially we can find ways to be more "forgiving" to give us more matches, while still not being _too_ forgiving so that we don't get images to match that really are different images.

Not sure that's a good idea. But worth trying.

Reply:
Per https://github.com/GoogleChromeLabs/wpp-research/pull/97#discussion_r1563203301: Given that (at least in my results) this is already over a million images, it's probably enough data so we don't need to go the lengths of making other assumptions about the loaded image URLs (which would also have risks).

Reply:
I was also thinking about how this could be made less strict

Does `performance.lcp_elem_stats.url` on the other hand include the query params?

As for the attributes, maybe I used the wrong one for `naturalWidth` and `naturalHeight` too. I thought `performance.lcp_elem_stats.naturalWidth/naturalHeight` were null, but maybe I was mistakenly checking `attributes.naturalWidth`. If it works, then I don't actually need to use `Images` data. Let me test again...